prboom2/src/gl_missingtexture.c

```diff
@@ -111,6 +159,56 @@ static void gld_PrepareSectorSpecialEffects(void)
   }
 }
 
+static void gld_RegisterBleedthroughSector(sector_t* source, sector_t* target, int ceiling)
+{
+    int i;
+    int source_idx = -1;
+    assert(source);
+    assert(target);
+
+    /* check whether the sector is processed already */
+    for (i = 0; i < numbleedsectors && source_idx == -1; i++)
+        if (bleedsectors[i].source == source && bleedsectors[i].ceiling == ceiling)
+            source_idx = i;
+    
+    if (source_idx == -1) {
+        /* allocate memory for new sector */
+        if (!bleedsectors) {
+            numbleedsectors = 1;
+            bleedsectors = (bleedthrough_t*) malloc(sizeof(bleedthrough_t));
+            memset(&bleedsectors[0], 0, sizeof(bleedthrough_t));
+        } else {
+            bleedsectors = (bleedthrough_t*) realloc(bleedsectors, (numbleedsectors + 1) * sizeof(bleedthrough_t));
+            memset(&bleedsectors[numbleedsectors], 0, sizeof(bleedthrough_t));
+            numbleedsectors++;
+        }
+
+        if(!bleedsectors) I_Error("gld_RegisterBleedthroughSector: Out of memory");
+
+        source_idx = numbleedsectors - 1;
+    }
+
+    bleedsectors[source_idx].source = source;
+    bleedsectors[source_idx].ceiling = ceiling;
+
+    /* either register the proposed target since it is first,
+     * or check if the new proposed target is a better option
+     * and register it instead */
+
+    if ((bleedsectors[source_idx].target == NULL) ||
+             (bleedsectors[source_idx].target && ((ceiling && bleedsectors[source_idx].target->ceilingheight > target->ceilingheight) || (bleedsectors[source_idx].target->floorheight < target->floorheight))))
+        bleedsectors[source_idx].target = target;
+}
+
+sector_t* GetBestBleedSector(sector_t* source, int ceiling)
+{
+    int i;
+    for (i = 0; i < numbleedsectors; i++)
+        if (bleedsectors[i].source == source && bleedsectors[i].ceiling == ceiling)
```

Comment:
Why only ceiling?

Reply:
Ceiling is a flag with encoding 0=floor, 1=ceiling; this is not my convention as I followed the existing `GetBestFake` implementation.

We could modify both to a better or more explicit indication, or I could use a different convention if it is preferred.